subprojects/core/src/main/java/org/gradle/api/services/internal/ConsumedBuildServiceProvider.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.services.internal;
+
+import org.gradle.api.artifacts.component.BuildIdentifier;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.api.services.BuildServiceRegistration;
+import org.gradle.api.services.BuildServiceRegistry;
+import org.gradle.internal.Cast;
+import org.gradle.internal.service.ServiceRegistry;
+
+import javax.annotation.Nullable;
+
+/**
+ * A build service that is consumed.
+ */
+public class ConsumedBuildServiceProvider<T extends BuildService<BuildServiceParameters>> extends BuildServiceProvider<T, BuildServiceParameters> {
+    protected final ServiceRegistry internalServices;
+    private final String serviceName;
+    private final Class<T> serviceType;
+    private final BuildIdentifier buildIdentifier;
+    private BuildServiceProvider<T, BuildServiceParameters> resolvedProvider;
+
+    public ConsumedBuildServiceProvider(
+        BuildIdentifier buildIdentifier,
+        String serviceName,
+        Class<T> serviceType,
+        ServiceRegistry internalServices
+    ) {
+        this.buildIdentifier = buildIdentifier;
+        this.serviceName = serviceName;
+        this.serviceType = serviceType;
+        this.internalServices = internalServices;
+    }
+
+    @Override
+    protected Value<? extends T> calculateOwnValue(ValueConsumer consumer) {
+        BuildServiceProvider<T, ?> resolvedProvider = resolve();
+        if (resolvedProvider == null) {
+            return Value.missing();
+        }
+        return resolvedProvider.calculateValue(consumer);
+    }
+
+    @Nullable
+    private BuildServiceProvider<T, BuildServiceParameters> resolve() {
```

Comment:
🧵 I assume this method could end up being called by multiple threads if references are shared between tasks?

Probably better to [make it thread-safe](https://github.com/gradle/gradle/pull/22696/commits/d67a4fdcb40a2ffbe4f698c968c07f5c49ec764f).

Reply:
I applied your changes, but it was thread-safe already, right? (in the worst case, it risked resolving multiple times - was that your concern?).

Reply:
`buildServiceRegistry.getRegistrations().findByName(serviceName)` was not thread-safe.

Reply:
uh, I missed that